From 5b5bd5d08d4be9e262d69529a946b7b4655a1731 Mon Sep 17 00:00:00 2001
From: Dan Crowell <dcrowell@us.ibm.com>
Date: Fri, 2 Oct 2015 14:33:22 -0500
Subject: [PATCH] Include MFSI offset for Centaur access from OCC

Add the FSI offset for the MFSI master to the base address for
cascaded slaves, this allows access to the Centaurs hanging off
the non-master processors from the OCC FIR collection code.

Change-Id: I5b6c1d77e90f233cab99bb74f5374d0456093cbc
CQ: SW322919
Reviewed-on: http://gfw160.aus.stglabs.ibm.com:8080/gerrit/20930
Tested-by: Jenkins Server
Tested-by: Jenkins OP Build CI
Tested-by: Jenkins OP HW
Tested-by: FSP CI Jenkins
Reviewed-by: Zane Shelley <zshelle@us.ibm.com>
Reviewed-by: MATTHEW A. PLOETZ <maploetz@us.ibm.com>
Reviewed-by: A. Patrick Williams III <iawillia@us.ibm.com>
---
 src/include/usr/fsi/fsiif.H | 2 +-
 src/usr/fsi/fsidd.C         | 8 ++++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/include/usr/fsi/fsiif.H b/src/include/usr/fsi/fsiif.H
index 5ab2e44..a161088 100644
--- a/src/include/usr/fsi/fsiif.H
+++ b/src/include/usr/fsi/fsiif.H
@@ -111,7 +111,7 @@ struct FsiLinkInfo_t
     uint8_t link; ///< Which link is this chip hanging off of
     uint8_t cascade; ///< Slave cascade position
     uint8_t mPort; ///< FSI Master port (0=A,1=B)
-    uint32_t baseAddr; ///< Base FSI Address for this chip
+    uint32_t baseAddr; ///< Base FSI Address for this chip, from master proc
 
     FsiLinkInfo_t() :
       master(NULL), type(TARGETING::FSI_MASTER_TYPE_NO_MASTER),
diff --git a/src/usr/fsi/fsidd.C b/src/usr/fsi/fsidd.C
index 6e7ae8a..f09363b 100644
--- a/src/usr/fsi/fsidd.C
+++ b/src/usr/fsi/fsidd.C
@@ -3031,6 +3031,14 @@ void FsiDD::getFsiLinkInfo( TARGETING::Target* i_slave,
     {
         o_info.mPort = 1;
     }
+
+    // if this chip is not off the master, need to include
+    //   its master's offset too
     o_info.baseAddr = addr_info.absAddr;
+    if( addr_info.opbTarg != iv_master )
+    {
+        FsiChipInfo_t mfsi_info = getFsiInfo(addr_info.accessInfo.master);
+        o_info.baseAddr |= getPortOffset(mfsi_info.type,mfsi_info.port);
+    }
 }
 
